core/http/client/src/main/java/org/eclipse/rdf4j/http/client/SPARQLProtocolSession.java

```diff
@@ -1139,6 +1150,35 @@ protected String getResponseMIMEType(HttpResponse method) throws IOException {
 		return null;
 	}
 
+	/**
+	 * Gets the character encoding specified in the HTTP headers of the supplied response, if any. For example, if the
+	 * response headers contain <tt>Content-Type: application/xml;charset=UTF-8</tt>, this method will return
+	 * {@link StandardCharsets#UTF_8 UTF-8} as the character encoding.
+	 * 
+	 * @param response the response to get the character encoding from.
+	 * @return the response character encoding, {@link Optional#empty()} if it can not be determined.
+	 */
+	Optional<Charset> getResponseCharset(HttpResponse response) {
+		Header[] headers = response.getHeaders("Content-Type");
```

Comment:
There are utilities in Http Client for obtaining the Charset! Something in the line of
```java
HttpEntity entity = response.getEntity();
ContentType contentType = ContentType.get(entity);
if (contentType == null) {
   return Optional.empty();
}
return Optional.ofNullable(contentType.getCharset());
```
Or is there a reason for not using it?

Reply:
I'm aware of the utilities, but I figured I'd do the minimal change only, so I just copied / adapted this from existing internal methods dealing with the content type. A lot of that code is from the httpclient 3.x days when it didn't have these kinds of utilities (or perhaps we weren't aware of it at the time). 

You're right that it could probably be cleaned up a bit though, I'll take another look at it.

Reply:
I've had a quick look but I'm not confident about changing how content-type header processing is handled in general without introducing regressions. It needs some better test coverage to get certainty on that and I just don't want to spend the effort right now (it's also somewhat out of scope for this ticket).  So I'll stick with the current approach for now. I have logged a separate ticket to deal with some general cleanup of this part of the code: #2619.